modules/sdk-coin-ada/src/ada.ts

```diff
@@ -207,6 +232,159 @@ export class Ada extends BaseCoin {
+      return res;
+    } catch (e) {
+      console.debug(e);
+    }
+    throw new Error(`Unable to call endpoint ${restEndpoint}`);
+  }
+
+  protected async getAddressInfo(
+    walletAddr: string
+  ): Promise<{ balance: number; utxoSet: Array<Record<string, any>> }> {
+    const requestBody = { _addresses: [walletAddr] };
+    const res = await this.getDataFromNode('address_info', requestBody);
+    if (res.status != 200) {
+      throw new Error(`Failed to retrieve address info for address ${walletAddr}`);
+    }
+    const body = res.body[0];
+    return { balance: body.balance, utxoSet: body.utxo_set };
+  }
+
+  protected async getChainTipInfo(): Promise<Record<string, string>> {
+    const res = await this.getDataFromNode('tip');
+    if (res.status != 200) {
+      throw new Error('Failed to retrieve chain tip info');
+    }
+    const body = res.body[0];
+    return body;
+  }
+
+  /**
+   * Builds a funds recovery transaction without BitGo
+   *
+   * @param {RecoveryOptions} params parameters needed to construct and
+   * (maybe) sign the transaction
+   *
+   * @returns {AdaTx} the serialized transaction hex string and index
+   * of the address being swept
+   */
+  async recover(params: RecoveryOptions): Promise<AdaTx> {
+    if (!params.bitgoKey) {
+      throw new Error('missing bitgoKey');
+    }
+    if (!params.recoveryDestination || !this.isValidAddress(params.recoveryDestination)) {
+      throw new Error('invalid recoveryDestination');
+    }
+    let addrFormat;
+    if (Ada.TestnetEnv.has(this.bitgo.getEnv())) {
+      addrFormat = AddressFormat.testnet;
+    } else if (Ada.MainnetEnv.has(this.bitgo.getEnv())) {
+      addrFormat = AddressFormat.mainnet;
+    } else {
+      throw new Error(`Could not derive the address type from environment ${this.bitgo.getEnv()}`);
+    }
+    let startIdx = params.startingScanIndex;
+    if (_.isUndefined(startIdx)) {
+      startIdx = 0;
+    } else if (!_.isInteger(startIdx) || startIdx < 0) {
+      throw new Error('Invalid starting index to scan for addresses');
+    }
+    let numIteration = params.scan;
+    if (_.isUndefined(numIteration)) {
+      numIteration = 20;
+    } else if (!_.isInteger(numIteration) || numIteration <= 0) {
+      throw new Error('Invalid scanning factor');
+    }
+    const bitgoKey = params.bitgoKey.replace(/\s/g, '');
+    const isUnsignedSweep = !params.userKey && !params.backupKey && !params.walletPassphrase;
+    const MPC = await EDDSAMethods.getInitializedMpcInstance();
+
+    for (let i = startIdx; i < numIteration + startIdx; i++) {
+      const currPath = `m/${i}`;
+      const accountId = MPC.deriveUnhardened(bitgoKey, currPath).slice(0, 64);
+      const keyPair = new AdaKeyPair({ pub: accountId });
+      const senderAddr = keyPair.getAddress(addrFormat);
+      const { balance, utxoSet } = await this.getAddressInfo(senderAddr);
+      if (balance <= 0) {
+        continue;
```

Comment:
do we need to log something when balance is 0 or negative?

Reply:
Joey didnt say anything about logging so i didnt. Maybe we should log an error if balance turns out to be strictly negative?

Reply:
no need to log imo, it'd be more for debugging purposes